lingpy/sequence/ngrams.py

```diff
@@ -0,0 +1,685 @@
+
+    # We generate the collection of ngrams for counting occurences
+    # with Python `zip()` function, so we can
+    # rely on internal C-code for speeding things up. What we do
+    # is build a list of arguments for the function by using a
+    # list comprehension over variable `order` and, then,
+    # decompose such list when passing it to the function.
+    # The list comprehension is built so that, for
+    # a sequence such as the characters in "Markov" (here without
+    # boundaries, but you should get the point) and an order of 3,
+    # we'll have:
+    #   [['M', 'a', 'r', 'k', 'o', 'v'],
+    #    ['a', 'r', 'k', 'o', 'v'],
+    #    ['r', 'k', 'o', 'v'],
+    #    ['k', 'o', 'v']]
+    #
+    # From which we zip all possible combinations.
+
+    for ngram in zip(*[seq[i:] for i in range(order)]):
+        yield ngram
+
+
+def get_all_ngrams(sequence, orders=None, pad_symbol='$'):
+    """
+    Build an iterator for collecting all ngrams of a sequence
+    of a given set of orders (i.e., "lengths"). If no set of
+    orders is provided, this will collect all possible ngrams
+    in the sequence.
+
+    Parameters
+    ----------
+    sequence: list or str
+        The sequence from which the ngrams will be collected.
+
+    orders: list
+        An optional list of the orders of the ngrams to
+        be collected. Can be larger than the length of the
+        sequence, in which case the latter will be padded
+        accordingly if requested. Defaults to the collection
+        of all possible ngrams in the sequence with the
+        minimum padding.
+
+    pad_symbol: object
+        An optional symbol to be used as start-of- and
+        end-of-sequence boundaries. The same symbol
+        is used for both boundaries. Must be a value
+        different from None, defaults to "$".
+
+    Returns
+    -------
+    out: iterable
+        An iterable over the ngrams of the sequence,
+        returned as tuples.
+
+    Examples
+    --------
+    >>> from lingpy.sequence import *
+    >>> sent = "Insurgents were killed"
+    >>> for ngram in get_all_ngrams(sent):
+    ...     print(ngram)
+    ...
+    ('Insurgents',)
+    ('were',)
+    ('killed',)
+    ('$', 'Insurgents')
+    ('Insurgents', 'were')
+    ('were', 'killed')
+    ('killed', '$')
+    ('$', '$', 'Insurgents')
+    ('$', 'Insurgents', 'were')
+    ('Insurgents', 'were', 'killed')
+    ('were', 'killed', '$')
+    ('killed', '$', '$')
+    """
+
+    # Convert to a tuple, for faster computation, compute the
+    # orders (if they were not given), and pad the sequence if
+    # requested.
+    seq = _seq_as_tuple(sequence)
+    if not orders:
+        orders = range(len(seq)+1)
+
+    for order in orders:
+        for ngram in get_n_ngrams(seq, order, pad_symbol):
+            yield ngram
+
+
+def get_skipngrams(sequence, order, max_gaps, pad_symbol='$', single_gap_opening=True):
+    """
+    Build an iterator for collecting all skip ngrams of a sequence
```

Comment:
I'd suggest to write this as `get_skip_ngrams`, to be more consistent.

Reply:
I think this structure is more consistent, I'm not sure if it is clear what each function does: ("traditional") ngrams, skip ngrams, and positional ngrams are fundamentally different in their usages, so the "grammar" I am using is something like: `get` + (optionally number/size) + type of ngrams. That's why we have:

- `get_n_ngrams` - traditional ngrams of (and only of) order `n`
- `get_all_ngrams` - traditional ngrams of all possible orders
- `get_skipngrams` - skip ngrams of an order `n` and number of gaps `g` (as it is not possible to define skip ngrams without these parameters, `get_ng_skipngrams` seemed redundant, but I can change it)
- `get_posngrams` - positional ngrams of a pre-order `p1` and a post-order `p2` (again, could be changed into `get_p_posngrams`)
- `get_all_posngrams` - positional ngrams of a given set of pre- and post-orders (this is the one which might actually be confusing, as it is not all *possible* positional ngrams, but all positional ngrams *up to* some specified orders)
- bigrams, trigrams, fourgrams - which could, for consistency (but without backwards compatibility) be renamed `get_2_ngrams` or `get_2grams` and so on.

What's our take?